app_dart/test/request_handlers/get_green_commits_test.dart

```diff
@@ -0,0 +1,113 @@
+import 'package:cocoon_service/src/service/datastore.dart';
+import 'package:gcloud/db.dart';
+import 'package:test/test.dart';
+
+import '../src/datastore/fake_config.dart';
+import '../src/request_handling/fake_authentication.dart';
+import '../src/request_handling/request_handler_tester.dart';
+import '../src/service/fake_build_status_provider.dart';
+import '../src/utilities/entity_generators.dart';
+
+void main() {
+  group('GetGreenCommits', () {
+    late FakeConfig config;
+    FakeClientContext clientContext;
+    FakeKeyHelper keyHelper;
+    FakeBuildStatusService buildStatusService;
+    late RequestHandlerTester tester;
+    late GetGreenCommits handler;
+
+    final Commit commit1 = generateCommit(1, timestamp: 3, sha: 'ea28a9c34dc701de891eaf74503ca4717019f829');
+    final Commit commit2 = generateCommit(2, timestamp: 1, sha: 'd5b0b3c8d1c5fd89302089077ccabbcfaae045e4');
+
+    final Task task1Succeed = generateTask(1, status: Task.statusSucceeded);
+    final Task task2Failed = generateTask(2, status: Task.statusFailed); // should fail if included
+    final Task task3FailedFlaky =
+        generateTask(3, status: Task.statusFailed, isFlaky: true); // should succeed if included because `bringup: true`
+    final Task task4SucceedFlaky = generateTask(4, status: Task.statusSucceeded, isFlaky: true);
+
+    final Stage stageOneSucceed =
+        Stage('cocoon', commit1, [task1Succeed], Task.statusInProgress); // should scceed, since task 1 succeed
+    final Stage stageFailed = Stage('luci', commit1, [task1Succeed, task2Failed],
+        Task.statusInProgress); // should fail, since task 1 succeed and task2 fail
+    final Stage stageMultipleSucceed = Stage('cocoon', commit2, [task1Succeed, task4SucceedFlaky],
+        Task.statusInProgress); // should succeed, since both task 1 and task 4 succeed
+    final Stage stageFailedFlaky = Stage('luci', commit2, [task1Succeed, task3FailedFlaky],
+        Task.statusInProgress); // should succeed, even though it includes task 3
+
+    Future<List<T?>?> decodeHandlerBody<T>() async {
+      final Body body = await tester.get(handler);
+      return (await utf8.decoder.bind(body.serialize() as Stream<List<int>>).transform(json.decoder).single
+              as List<dynamic>)
+          .cast<T>();
+    }
+
+    setUp(() {
+      clientContext = FakeClientContext();
+      keyHelper = FakeKeyHelper(applicationContext: clientContext.applicationContext);
+      tester = RequestHandlerTester();
+      config = FakeConfig(keyHelperValue: keyHelper);
+      buildStatusService = FakeBuildStatusService(commitStatuses: <CommitStatus>[]);
+      handler = GetGreenCommits(
+        config,
+        datastoreProvider: (DatastoreDB db) => DatastoreService(config.db, 5),
+        buildStatusProvider: (_) => buildStatusService,
+      );
+    });
+
+    test('no green commits', () async {
+      final List<String?> result = (await decodeHandlerBody())!;
+      expect(result, isEmpty);
+    });
+
+    test(
```

Comment:
It's best practice to keep your tests small, and only validate a specific aspect. For this, you can tell this test is doing too much by having "and" and "but" mulitple times in the description. In these cases, either split into smaller tests or wrap in ` group`